File: src/utils/mealPlanner.js
/**
 * "Plan my week" suggestion engine. Pure and deterministic given `random`, so it
 * can be unit-tested; the iOS port in MealPlanner.swift mirrors it rule for rule.
 *
 * Variety comes from:
 *  - a per-recipe rest period scaled to how often that recipe is usually made,
 *  - weekly caps on the same protein / cuisine (attribute-level variety),
 *  - one "something new" slot per week,
 *  - resurfacing favourites that haven't been made in a long while,
 *  - quick meals on weeknights, bigger cooks at the weekend,
 *  - nudging recipes that share fresh ingredients onto nearby days,
 *  - a little randomness so equal candidates don't always win in the same order.
 */

const DAY_MS = 24 * 60 * 60 * 1000;
const DEFAULT_GAP_DAYS = 21;
const WEEKDAY_NAMES = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
const MAIN_COURSES = new Set([null, undefined, 'main']);

const keyToUtc = (key) => {
  const [y, m, d] = key.split('-').map(Number);
  return Date.UTC(y, m - 1, d);
};

/** Whole days from date key `a` to date key `b`. */
export const daysBetween = (a, b) => Math.round((keyToUtc(b) - keyToUtc(a)) / DAY_MS);

/** Monday = 0 … Sunday = 6 for a date key. */
export const weekdayIndex = (key) => (new Date(keyToUtc(key)).getUTCDay() + 6) % 7;

/**
 * Typical days between cooks: the median gap once a recipe has been made twice,
 * clamped to a sensible range; three weeks before that.
 * @param {string[]} cookKeys - Date keys of completed cooks, any order
 */
export const typicalGapDays = (cookKeys) => {
  if (!cookKeys || cookKeys.length < 2) return DEFAULT_GAP_DAYS;
  const sorted = [...cookKeys].sort();
  const gaps = [];
  for (let i = 1; i < sorted.length; i += 1) {
    const gap = daysBetween(sorted[i - 1], sorted[i]);
    if (gap > 0) gaps.push(gap);
  }
  if (gaps.length === 0) return DEFAULT_GAP_DAYS;
  gaps.sort((a, b) => a - b);
  const mid = Math.floor(gaps.length / 2);
  const median = gaps.length % 2 ? gaps[mid] : (gaps[mid - 1] + gaps[mid]) / 2;
  return Math.min(90, Math.max(7, median));
};

const isEligibleForMeal = (attrs, meal) => {
  if (attrs && !MAIN_COURSES.has(attrs.course) && !(meal === 'breakfast' && attrs.course === 'snack')) return false;
  if (attrs?.mealTypes?.length) return attrs.mealTypes.includes(meal);
  // Untagged (or no meals set): fine for lunch/dinner at reduced confidence, never breakfast.
  return meal !== 'breakfast';
};

const hasMealTag = (attrs, meal) => Boolean(attrs?.mealTypes?.includes(meal));

const weeksText = (days) => {
  const weeks = Math.round(days / 7);
  return weeks <= 1 ? 'over a week' : `${weeks} weeks`;
};

/**
 * Builds suggestions for the empty slots of one week.
 *
 * @param {object} input
 * @param {Array<{date: string, meal: string}>} input.slots - Empty, active slots in day→meal order
 * @param {Array<{date: string, meal: string, recipeId: string|null}>} input.filled - Slots already
 *   holding something this week (manual, kept or locked); their recipes count toward variety
 * @param {Array<{id: string, name: string, cookCount?: number, lastCookedAt?: string|null}>} input.recipes
 * @param {Map<string, object>} input.attributes - recipeId → attributes (camelCase)
 * @param {Map<string, string[]>} input.cookDates - recipeId → date keys of completed cooks
 * @param {Map<string, string>} input.lastPlanned - recipeId → latest date key it was planned before this week
 * @param {Map<string, Set<string>>} [input.excluded] - `date|meal` → recipe ids not to suggest there (swaps)
 * @param {() => number} [input.random] - Returns [0, 1); injectable for tests
 * @returns {{suggestions: Array<{date: string, meal: string, recipeId: string, reason: string}>, libraryNote: string|null}}
 */
export const planWeek = ({
  slots,
  filled = [],
  recipes,
  attributes,
  cookDates,
  lastPlanned,
  excluded = new Map(),
  random = Math.random,
}) => {
  const chosen = filled
    .filter((f) => f.recipeId)
    .map((f) => ({ date: f.date, meal: f.meal, recipeId: f.recipeId }));
  const suggestions = [];
  let usedFallback = false;

  const profile = new Map(recipes.map((r) => {
    const cooks = cookDates.get(r.id) ?? [];
    const lastCook = r.lastCookedAt ? r.lastCookedAt.slice(0, 10) : cooks.slice().sort().at(-1) ?? null;
    const planned = lastPlanned.get(r.id) ?? null;
    const lastSeen = [lastCook, planned].filter(Boolean).sort().at(-1) ?? null;
    return [r.id, {
      recipe: r,
      attrs: attributes.get(r.id) ?? null,
      gap: typicalGapDays(cooks),
      cookCount: r.cookCount ?? cooks.length,
      lastSeen,
      isNew: !lastCook && !planned && (r.cookCount ?? 0) === 0,
    }];
  }));

  // One "something new" slot when planning a week: the first empty dinner from Wednesday on,
  // else any empty dinner. A single-slot swap never claims it.
  const exploreSlot = slots.length < 2 ? null
    : slots.find((s) => s.meal === 'dinner' && weekdayIndex(s.date) >= 2)
      ?? slots.find((s) => s.meal === 'dinner') ?? null;

  const scoreFor = (p, slot, relaxed) => {
    const { attrs } = p;
    if (!isEligibleForMeal(attrs, slot.meal)) return null;
    const blocked = excluded.get(`${slot.date}|${slot.meal}`);
    if (blocked?.has(p.recipe.id)) return null;

    const weekChosen = chosen.map((c) => ({ ...c, p: profile.get(c.recipeId) })).filter((c) => c.p);
    const alreadyThisWeek = weekChosen.some((c) => c.recipeId === p.recipe.id);
    const sameDay = weekChosen.some((c) => c.recipeId === p.recipe.id && c.date === slot.date);
    if (sameDay || (alreadyThisWeek && !relaxed)) return null;

    const daysSince = p.lastSeen ? daysBetween(p.lastSeen, slot.date) : null;
    const resting = daysSince !== null && daysSince < 0.6 * p.gap;
    if (resting && !relaxed) return null;

    const protein = attrs?.protein;
    const cuisine = attrs?.cuisine;
    const proteinCount = protein && protein !== 'none'
      ? weekChosen.filter((c) => c.p.attrs?.protein === protein).length : 0;
    const cuisineCount = cuisine && cuisine !== 'other'
      ? weekChosen.filter((c) => c.p.attrs?.cuisine === cuisine).length : 0;
    if (!relaxed && (proteinCount >= 2 || cuisineCount >= 3)) return null;

    let score = hasMealTag(attrs, slot.meal) ? 1 : 0.7;
    const reasons = [];

    score *= daysSince === null ? 1 : Math.min(1, daysSince / p.gap);
    score *= 1 + 0.25 * Math.log1p(p.cookCount);

    if (p.cookCount >= 2 && daysSince !== null && daysSince >= 2 * p.gap) {
      score *= 1.35;
      reasons.push({ weight: 3, text: `Haven't made this in ${weeksText(daysSince)}` });
    }

    if (p.isNew) {
      const isExplore = exploreSlot && slot.date === exploreSlot.date && slot.meal === exploreSlot.meal;
      score *= isExplore ? 2.5 : 0.85;
      if (isExplore) reasons.push({ weight: 4, text: 'Something new to try' });
    }

    score *= 0.6 ** proteinCount;
    score *= 0.75 ** cuisineCount;

    const weekday = weekdayIndex(slot.date);
    const isWeeknight = weekday <= 3 && slot.meal !== 'breakfast';
    if (attrs?.effort === 'project') score *= isWeeknight ? 0.4 : (weekday >= 5 ? 1.1 : 1);
    if (attrs?.effort === 'quick' && isWeeknight) {
      score *= 1.15;
      reasons.push({ weight: 1, text: 'Quick for a weeknight' });
    }

    if (attrs?.method) {
      const adjacentSameMethod = weekChosen.some((c) => c.meal === slot.meal
        && Math.abs(daysBetween(c.date, slot.date)) === 1 && c.p.attrs?.method === attrs.method);
      if (adjacentSameMethod) score *= 0.85;
    }

    const fresh = attrs?.perishables ?? [];
    if (fresh.length) {
      const shared = [];
      for (const c of weekChosen) {
        const distance = daysBetween(c.date, slot.date);
        if (distance < 0 || distance > 3) continue;
        for (const item of c.p.attrs?.perishables ?? []) {
          if (fresh.includes(item) && !shared.some((s) => s.item === item)) shared.push({ item, date: c.date });
        }
      }
      if (shared.length) {
        score *= 1 + 0.15 * Math.min(shared.length, 2);
        reasons.push({ weight: 2, text: `Uses the rest of ${WEEKDAY_NAMES[weekdayIndex(shared[0].date)]}'s ${shared[0].item}` });
      }
    }

    if (p.cookCount >= 3) reasons.push({ weight: 0.5, text: `A regular — made ${p.cookCount} times` });
    if (daysSince !== null && daysSince >= 14) reasons.push({ weight: 0.25, text: `Last made ${weeksText(daysSince)} ago` });

    if (relaxed) score *= 0.3;
    score *= 0.85 + 0.3 * random();

    const reason = reasons.sort((a, b) => b.weight - a.weight)[0]?.text ?? `Good for ${slot.meal}`;
    return { score, reason };
  };

  for (const slot of slots) {
    let best = null;
    for (const relaxed of [false, true]) {
      for (const p of profile.values()) {
        const result = scoreFor(p, slot, relaxed);
        if (result && (!best || result.score > best.score)) best = { ...result, recipeId: p.recipe.id };
      }
      if (best) {
        if (relaxed) usedFallback = true;
        break;
      }
    }
    if (!best) continue;
    const suggestion = { date: slot.date, meal: slot.meal, recipeId: best.recipeId, reason: best.reason };
    suggestions.push(suggestion);
    chosen.push(suggestion);
  }

  const dinnerLibrary = [...profile.values()].filter((p) => isEligibleForMeal(p.attrs, 'dinner')).length;
  let libraryNote = null;
  if (dinnerLibrary === 0) {
    libraryNote = 'Add a few recipes to get suggestions.';
  } else if (usedFallback || dinnerLibrary < 8) {
    libraryNote = `You have ${dinnerLibrary} ${dinnerLibrary === 1 ? 'recipe' : 'recipes'} for main meals — enough for about ${Math.max(1, Math.floor(dinnerLibrary / 5))} week${dinnerLibrary >= 10 ? 's' : ''} without repeats. Add more and the plan will repeat less.`;
  }

  return { suggestions, libraryNote };
};

File: ios-native/GatherLists/GatherLists/Services/MealPlanner.swift
import Foundation

/// "Plan my week" suggestion engine. A rule-for-rule port of `src/utils/mealPlanner.js`
/// (which has the unit tests); keep the two in step.
///
/// Variety comes from a per-recipe rest period scaled to how often it's usually made, weekly
/// caps on the same protein / cuisine, one "something new" slot, resurfacing long-forgotten
/// favourites, quick meals on weeknights, nudging recipes that share fresh ingredients onto
/// nearby days, and learned household preference (Thompson-style: kept/cooked vs swapped/skipped,
/// fading over time), which also supplies the randomness between equal candidates.
enum MealPlanner {
    struct Slot: Hashable {
        let date: String
        let meal: MealType
        var key: String { "\(date)|\(meal.rawValue)" }
    }

    struct Filled {
        let date: String
        let meal: MealType
        let recipeId: UUID?
    }

    struct Suggestion {
        let date: String
        let meal: MealType
        let recipeId: UUID
        let reason: String
    }

    struct Input {
        var slots: [Slot]
        var filled: [Filled]
        var recipes: [Recipe]
        var attributes: [UUID: RecipeAttributes]
        /// Date keys of completed cooks per recipe.
        var cookDates: [UUID: [String]]
        /// Latest date key each recipe was planned before this week.
        var lastPlanned: [UUID: String]
        /// Recipe ids not to suggest per slot key (swaps / regenerate).
        var excluded: [String: Set<UUID>] = [:]
        var preferences: [UUID: Preference] = [:]
        /// Weekdays that need quick meals (Monday = 0), from `learnQuickWeekdays`.
        var quickWeekdays: Set<Int> = MealPlanner.defaultQuickWeekdays
        var random: () -> Double = { Double.random(in: 0..<1) }
    }

    /// Learned liking for a recipe: decayed weights of positive and negative reactions.
    struct Preference {
        var positive: Double = 0
        var negative: Double = 0
    }

    enum FeedbackEvent: String {
        case kept, swapped, regenerated
    }

    struct Result {
        let suggestions: [Suggestion]
        let libraryNote: String?
    }

    private static let defaultGapDays = 21.0
    static let defaultQuickWeekdays: Set<Int> = [0, 1, 2, 3]
    private static let feedbackHalfLifeDays = 60.0
    private static let weekdayNames = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]

    private struct Profile {
        let recipe: Recipe
        let attrs: RecipeAttributes?
        let gap: Double
        let cookCount: Int
        let lastSeen: String?
        let isNew: Bool
    }

    private struct Reason {
        let weight: Double
        let text: String
    }

    // MARK: - Date helpers

    private static let utcCalendar: Calendar = {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        return calendar
    }()

    private static func dayNumber(_ key: String) -> Int {
        let parts = key.split(separator: "-").compactMap { Int($0) }
        guard parts.count == 3,
              let date = utcCalendar.date(from: DateComponents(year: parts[0], month: parts[1], day: parts[2]))
        else { return 0 }
        return Int((date.timeIntervalSince1970 / 86_400).rounded())
    }

    /// Whole days from date key `a` to date key `b`.
    static func daysBetween(_ a: String, _ b: String) -> Int {
        dayNumber(b) - dayNumber(a)
    }

    /// Monday = 0 … Sunday = 6.
    static func weekdayIndex(_ key: String) -> Int {
        // 1970-01-01 was a Thursday (index 3).
        ((dayNumber(key) + 3) % 7 + 7) % 7
    }

    /// Median days between cooks once made twice, clamped to 7…90; three weeks before that.
    static func typicalGapDays(_ cookKeys: [String]) -> Double {
        guard cookKeys.count >= 2 else { return defaultGapDays }
        let sorted = cookKeys.sorted()
        let gaps = zip(sorted, sorted.dropFirst()).map { daysBetween($0, $1) }.filter { $0 > 0 }.sorted()
        guard !gaps.isEmpty else { return defaultGapDays }
        let mid = gaps.count / 2
        let median = gaps.count % 2 == 1 ? Double(gaps[mid]) : Double(gaps[mid - 1] + gaps[mid]) / 2
        return min(90, max(7, median))
    }

    // MARK: - Learning

    /// Builds each recipe's learned preference; a reaction loses half its weight every 60 days.
    static func buildPreferences(
        todayKey: String,
        cooks: [(recipeId: UUID, date: String)],
        feedback: [(recipeId: UUID, event: FeedbackEvent, date: String)],
        planned: [(recipeId: UUID, date: String, cooked: Bool)]
    ) -> [UUID: Preference] {
        var prefs: [UUID: Preference] = [:]
        func add(_ recipeId: UUID, _ date: String, _ weight: Double) {
            let age = Double(max(0, daysBetween(date, todayKey)))
            let decayed = weight * pow(0.5, age / feedbackHalfLifeDays)
            if decayed > 0 {
                prefs[recipeId, default: Preference()].positive += decayed
            } else {
                prefs[recipeId, default: Preference()].negative -= decayed
            }
        }
        for cook in cooks { add(cook.recipeId, cook.date, 1) }
        for item in feedback {
            switch item.event {
            case .kept: add(item.recipeId, item.date, 0.75)
            case .swapped: add(item.recipeId, item.date, -1)
            case .regenerated: add(item.recipeId, item.date, -0.5)
            }
        }
        for item in planned where !item.cooked && item.date < todayKey {
            add(item.recipeId, item.date, -0.3)
        }
        return prefs
    }

    /// Multiplier in 0.7…1.3 sampled around the Beta(1+liked, 1+disliked) mean, spread by its uncertainty.
    private static func preferenceFactor(_ pref: Preference?, random: () -> Double) -> Double {
        let a = 1 + (pref?.positive ?? 0)
        let b = 1 + (pref?.negative ?? 0)
        let mean = a / (a + b)
        let sd = sqrt((a * b) / (pow(a + b, 2) * (a + b + 1)))
        let sample = min(1, max(0, mean + (random() - 0.5) * 3.4 * sd))
        return 0.7 + 0.6 * sample
    }

    /// Learns busy weekdays from cook durations: with 8+ cooks, a day is quick if its median cook
    /// is 40 minutes or less, or it's a weekday (Mon–Fri) the household rarely cooks on.
    static func learnQuickWeekdays(_ sessions: [(weekday: Int, minutes: Double)]) -> Set<Int> {
        guard sessions.count >= 8 else { return defaultQuickWeekdays }
        var quick: Set<Int> = []
        for day in 0..<7 {
            let durations = sessions.filter { $0.weekday == day }.map(\.minutes).sorted()
            if durations.count <= 1 {
                if day <= 4 { quick.insert(day) }
                continue
            }
            if durations[durations.count / 2] <= 40 { quick.insert(day) }
        }
        return quick
    }

    private static func isEligible(_ attrs: RecipeAttributes?, for meal: MealType) -> Bool {
        if let attrs {
            let course = attrs.course
            let isMain = course == nil || course == RecipeCourse.main.rawValue
            if !isMain && !(meal == .breakfast && course == RecipeCourse.snack.rawValue) { return false }
            if !attrs.mealTypes.isEmpty { return attrs.mealTypes.contains(meal.rawValue) }
        }
        return meal != .breakfast
    }

    private static func weeksText(_ days: Int) -> String {
        let weeks = Int((Double(days) / 7).rounded())
        return weeks <= 1 ? "over a week" : "\(weeks) weeks"
    }

    // MARK: - Planning

    static func planWeek(_ input: Input) -> Result {
        var chosen: [Filled] = input.filled.filter { $0.recipeId != nil }
        var suggestions: [Suggestion] = []
        var usedFallback = false

        var profiles: [UUID: Profile] = [:]
        for recipe in input.recipes {
            let cooks = input.cookDates[recipe.id] ?? []
            let lastCook = recipe.lastCookedAt.map { MealPlanWeek.key(for: $0) } ?? cooks.max()
            let planned = input.lastPlanned[recipe.id]
            let lastSeen = [lastCook, planned].compactMap { $0 }.max()
            profiles[recipe.id] = Profile(
                recipe: recipe,
                attrs: input.attributes[recipe.id],
                gap: typicalGapDays(cooks),
                cookCount: recipe.cookCount ?? cooks.count,
                lastSeen: lastSeen,
                isNew: lastCook == nil && planned == nil && (recipe.cookCount ?? 0) == 0
            )
        }
        // Stable iteration order, matching the JS Map insertion order (recipes array order).
        let ordered = input.recipes.compactMap { profiles[$0.id] }

        // A single-slot swap never claims the week's "something new" slot.
        let exploreSlot = input.slots.count < 2 ? nil
            : input.slots.first { $0.meal == .dinner && weekdayIndex($0.date) >= 2 }
                ?? input.slots.first { $0.meal == .dinner }

        func score(_ p: Profile, _ slot: Slot, relaxed: Bool) -> (score: Double, reason: String)? {
            let attrs = p.attrs
            guard isEligible(attrs, for: slot.meal) else { return nil }
            if input.excluded[slot.key]?.contains(p.recipe.id) == true { return nil }

            let weekChosen = chosen.compactMap { c -> (Filled, Profile)? in
                guard let id = c.recipeId, let profile = profiles[id] else { return nil }
                return (c, profile)
            }
            let alreadyThisWeek = weekChosen.contains { $0.0.recipeId == p.recipe.id }
            let sameDay = weekChosen.contains { $0.0.recipeId == p.recipe.id && $0.0.date == slot.date }
            if sameDay || (alreadyThisWeek && !relaxed) { return nil }

            let daysSince = p.lastSeen.map { daysBetween($0, slot.date) }
            if let daysSince, Double(daysSince) < 0.6 * p.gap, !relaxed { return nil }

            let protein = attrs?.protein
            let cuisine = attrs?.cuisine
            let proteinCount = (protein != nil && protein != RecipeProtein.none.rawValue)
                ? weekChosen.filter { $0.1.attrs?.protein == protein }.count : 0
            let cuisineCount = (cuisine != nil && cuisine != RecipeCuisine.other.rawValue)
                ? weekChosen.filter { $0.1.attrs?.cuisine == cuisine }.count : 0
            if !relaxed && (proteinCount >= 2 || cuisineCount >= 3) { return nil }

            var value = attrs?.mealTypes.contains(slot.meal.rawValue) == true ? 1.0 : 0.7
            var reasons: [Reason] = []

            value *= daysSince.map { min(1, Double($0) / p.gap) } ?? 1
            value *= 1 + 0.25 * log1p(Double(p.cookCount))

            if p.cookCount >= 2, let daysSince, Double(daysSince) >= 2 * p.gap {
                value *= 1.35
                reasons.append(Reason(weight: 3, text: "Haven't made this in \(weeksText(daysSince))"))
            }

            if p.isNew {
                let isExplore = exploreSlot == slot
                value *= isExplore ? 2.5 : 0.85
                if isExplore { reasons.append(Reason(weight: 4, text: "Something new to try")) }
            }

            value *= pow(0.6, Double(proteinCount))
            value *= pow(0.75, Double(cuisineCount))

            let weekday = weekdayIndex(slot.date)
            let isBusyDay = input.quickWeekdays.contains(weekday) && slot.meal != .breakfast
            if attrs?.effort == RecipeEffort.project.rawValue {
                value *= isBusyDay ? 0.4 : (weekday >= 5 ? 1.1 : 1)
            }
            if attrs?.effort == RecipeEffort.quick.rawValue && isBusyDay {
                value *= 1.15
                reasons.append(Reason(weight: 1, text: "Quick for a busy day"))
            }

            if let method = attrs?.method {
                let adjacentSameMethod = weekChosen.contains {
                    $0.0.meal == slot.meal && abs(daysBetween($0.0.date, slot.date)) == 1 && $0.1.attrs?.method == method
                }
                if adjacentSameMethod { value *= 0.85 }
            }

            let fresh = attrs?.perishables ?? []
            if !fresh.isEmpty {
                var shared: [(item: String, date: String)] = []
                for (c, profile) in weekChosen {
                    let distance = daysBetween(c.date, slot.date)
                    guard distance >= 0, distance <= 3 else { continue }
                    for item in profile.attrs?.perishables ?? [] where fresh.contains(item) && !shared.contains(where: { $0.item == item }) {
                        shared.append((item, c.date))
                    }
                }
                if let first = shared.first {
                    value *= 1 + 0.15 * Double(min(shared.count, 2))
                    reasons.append(Reason(weight: 2, text: "Uses the rest of \(weekdayNames[weekdayIndex(first.date)])'s \(first.item)"))
                }
            }

            let pref = input.preferences[p.recipe.id]
            if let pref, pref.positive >= 2, (1 + pref.positive) / (2 + pref.positive + pref.negative) >= 0.75 {
                reasons.append(Reason(weight: 1.5, text: "A household favourite"))
            }
            if p.cookCount >= 3 { reasons.append(Reason(weight: 0.5, text: "A regular — made \(p.cookCount) times")) }
            if let daysSince, daysSince >= 14 { reasons.append(Reason(weight: 0.25, text: "Last made \(weeksText(daysSince)) ago")) }

            if relaxed { value *= 0.3 }
            value *= preferenceFactor(pref, random: input.random)

            let reason = reasons.max { $0.weight < $1.weight }?.text ?? "Good for \(slot.meal.rawValue)"
            return (value, reason)
        }

        for slot in input.slots {
            var best: (score: Double, reason: String, recipeId: UUID)?
            for relaxed in [false, true] {
                for p in ordered {
                    if let result = score(p, slot, relaxed: relaxed), result.score > (best?.score ?? -1) {
                        best = (result.score, result.reason, p.recipe.id)
                    }
                }
                if best != nil {
                    if relaxed { usedFallback = true }
                    break
                }
            }
            guard let best else { continue }
            suggestions.append(Suggestion(date: slot.date, meal: slot.meal, recipeId: best.recipeId, reason: best.reason))
            chosen.append(Filled(date: slot.date, meal: slot.meal, recipeId: best.recipeId))
        }

        let library = ordered.filter { isEligible($0.attrs, for: .dinner) }.count
        var note: String?
        if library == 0 {
            note = "Add a few recipes to get suggestions."
        } else if usedFallback || library < 8 {
            let weeks = max(1, library / 5)
            note = "You have \(library) \(library == 1 ? "recipe" : "recipes") for main meals — enough for about \(weeks) week\(library >= 10 ? "s" : "") without repeats. Add more and the plan will repeat less."
        }
        return Result(suggestions: suggestions, libraryNote: note)
    }
}
